File: archive/lc_5. Longest Palindromic Substring.java
// String, Dynamic Programming (d)
// Manacher's Algorithm

class Solution {
    public String longestPalindrome(String s) {
        char[] cs = new char[s.length()*2+1];
        for (int i=0, l=s.length(); i<l; ++i) {
            cs[i*2] = '#';
            cs[i*2+1] = s.charAt(i);
        }
        cs[cs.length-1] = '#';
        int [] d = new int[cs.length];
        int pre = 0, mxi = 0;
        for (int i=1, l=cs.length; i<l; ++i) {
            if (i <= pre + d[pre]) {
                int mi = pre - (i - pre);
                if (i + d[mi] != pre + d[pre]) {
                    d[i] = Math.min(d[mi], pre + d[pre] - i);
                } else {
                    d[i] = d[mi];
                    while (i-d[i]>0 && i+d[i]+1<l && cs[i-d[i]-1] == cs[i+d[i]+1]) ++d[i];
                }
            } else {
                while (i-d[i]>0 && i+d[i]+1<l && cs[i-d[i]-1] == cs[i+d[i]+1]) ++d[i];
            }
            if (i + d[i] > pre + d[pre]) pre = i;
            if (d[i] > d[mxi]) mxi = i;
        }
        return s.substring((mxi-d[mxi])/2, (mxi+d[mxi])/2);
    }
}

File: archive/lc_5. Longest Palindromic Substring.py
# String, Dynamic Programming (dp)
# Manacher's Algorithm

class Solution:
    def longestPalindrome(self, s: str) -> str:
        s2 = '#' + '#'.join(s) + '#'
        ln2 = len(s2)
        d = [1] * ln2
        pre = 0
        for i in range(1, ln2 - 1):
            if pre + d[pre] > i:
                d[i] = min(pre + d[pre] - i, d[pre-(i-pre)])
            while 0 <= i-d[i] and i+d[i] < ln2 and s2[i-d[i]] == s2[i+d[i]]:
                d[i] += 1
            if i + d[i] > pre + d[pre]: pre = i
        d = [x-1 for x in d]
        mxi = max(range(ln2), key=d.__getitem__)
        return s[(mxi-d[mxi])//2:(mxi+d[mxi])//2]